docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+* Windows: stored in `"C:\Program Files\machineadm\plugins"`
+
+### Security Model
+
+#### Userdata storage
+
+Machineadm's security model is intended such that secrets associated with machine configuration are stored on the
+management Kubernetes cluster and are secured to the level that the backing API server and etcd are secured.
+
+For machine bootstrapping, secrets are intended to be delivered by an infrastructure provider to a suitable secure
+location, and to also provide a mechanism for the machine to report back status. Examples include:
+
+* Amazon S3 or S3-compatible API such as Minio
+* AWS Secrets Manager and AWS Systems Manager Parameter Store
+
+When using external storage, we assume security is configured appropriately (e.g. Amazon Identity and Access Management),
+and is outside of the system boundary of machineadm bootstrap controller.
+
+In addition, some providers may require a secondary mechanism to encrypt data. For example, a deployment of an S3 compatible
+API may not support the same levels of fine-grained IAM access supported by EC2. A provider may then want to further encrypt
+the data. In this case  we will additionally support a data block for in-band binary data (see [#encrypted-configuration](#encrypted-configuration)) that can be decrypted with a key derivation algorithm and a passphrase from an external source. We do not assume a KMS can
+directly encrypt the entire machine configuration as most systems have byte limits of around 64KB. Instead we assume
+the use of envelope encryption where a KMS encrypts a random string that is itself used as a key for encryption/data
+of data.
+
+We furthermore protect the KMS issued random string by deriving a key from it using SHA-512 PBKDF2-HMAC hashed to 50,000 rounds, with the result used for AES-256-GCM. This provides FIPS compliance with a suitable level of encryption given the
+resultant data will be applied as plain text to the machine configuration.
+
+The passphrase may either be stored at a remote location only the machine can access, or itself may be encrypted by an external Key Management Service, in which case machineadm is using envelope encryption. For example. the AWS provider could call `kms::GenerateDataKey` that generates a random string and associated ciphertext for the string encrypted by the Customer Managed Key. The provider could then use the random string as the passphrase for key derivation and throw away the plaintext
+passphrase, encoding the KMS encrypted value in a URI understood by the machineadm plugin, e.g. `kms://ZW5jcnlwdGVk@arn:aws:kms:us-west-2:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab`, where the base64 value before
+@ represents the encrypted passphrase to be handed to kms::Decrypt.
+
+In terms of why we provide support for both external data and in-band data with encryption is that some infrastructure
+APIs have limited userdata storage: For example, AWS and Azure are both limited to 64-bytes for userdata. When users
```

Comment:
```suggestion
APIs have limited userdata storage: For example, AWS and Azure are both limited to 64 kilobytes for userdata. When users
```

nit

Reply:
LOL, yup. 64-bytes would be particularly challenging.